crates/wasmtime/src/store.rs

```diff
@@ -1669,6 +1734,26 @@ unsafe impl<T> wasmtime_runtime::Store for StoreInner<T> {
 
         impl std::error::Error for OutOfGasError {}
     }
+
+    #[cfg(feature = "async")]
+    fn new_epoch(&mut self) {
+        // Yield unconditionally.
+        let _ = self.out_of_gas_yield(0);
+        // Set a new deadline.
+        self.reset_epoch_deadline(self.epoch_deadline_refill_delta);
+    }
+
+    #[cfg(not(feature = "async"))]
+    fn new_epoch(&mut self) {}
+}
+
+impl<T> StoreInner<T> {
+    #[cfg(feature = "async")]
+    pub(crate) fn reset_epoch_deadline(&mut self, delta: u64) {
+        // Set a new deadline based on the "epoch deadline delta".
```

Comment:
Could this comment how the deadline is store-local and not operated on by other threads so the `unsafe` here should be ok? Also that this value is cached by compiled code on the stack but it's reloaded before any new epoch is entered.

Reply:
Added a comment describing the safety here.